### ISSUE TITLE
AsyncTask list can orphan the  the queue 

### ISSUE THREAD

**Issue by Narimm** [MockBukkit/MockBukkit#162]
```
[executorService] executorService = {BukkitSchedulerMock@5945} 
   currentTick = 5645936
   id = 15
   tasks = {LinkedList@6053}  size = 0
   pool = {ThreadPoolExecutor@5948} "java.util.concurrent.ThreadPoolExecutor@1cd2143b[Running, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 6]"
   asyncTasksRunning = {AtomicInteger@5949} "0"
   value = 0
   asyncException = {AtomicReference@5950} "be.seeseemelk.mockbukkit.UnimplementedOperationException: Not implemented"
   value = {UnimplementedOperationException@6026} "be.seeseemelk.mockbukkit.UnimplementedOperationException: Not implemented"
   asyncTasksQueued = 2
```
 
See above that there are 2 tasks queued but the task list is empty....clearly this will never complete.
I think this state exists if a scheduledasynctask is cancelled before it ever runs.  This method needs a safety check if this occurs

```java
public void performOneTick()
	{
		currentTick++;
		List<ScheduledTask> oldTasks = tasks;
		tasks = new LinkedList<>();

		for (ScheduledTask task : oldTasks)
		{
			if (task.getScheduledTick() == currentTick && !task.isCancelled())
			{
				if (task.isSync())
				{
					task.run();
				}
				else
				{
					asyncTasksRunning.incrementAndGet();
					pool.execute(task.getRunnable());
					asyncTasksQueued--;
				}

				if (task instanceof RepeatingTask && !task.isCancelled())
				{
					((RepeatingTask) task).updateScheduledTick();
					tasks.add(task);
				}
			}
			else if (!task.isCancelled())
			{
				tasks.add(task);
			}
		}
	}
```

**Comment by Narimm** [comment]
Just to be clear this results in the  `BukkitSchedulerMock.waitAsyncTasksFinished()` never returning.

**Comment by Narimm** [comment]
I came up with a few options to handle this it primarily blocks ` MockBukkit.unmock() `
1. Force a shutdown.
2. Reset the integer.
3. Limit the time you wait for async tasks to run

adding this to the tick would probably be enough

```java
		if (oldTasks.size() == 0 && asyncTasksQueued > 0) {
			Logger.getLogger(LOGGER_NAME).warning("Asynchronous Tasks may have been lost -> " + asyncTasksQueued);
			asyncTasksQueued = 0;
		}```

**Comment by gabber235** [comment]
For me this is happening when I cancel the task since the cancel operation does nothing with the `asyncTasksQueued`

```java
	@Override
	public void cancelTask(int taskId)
	{
		for (ScheduledTask task : tasks)
		{
			if (task.getTaskId() == taskId)
			{
				task.cancel();
				return;
			}
		}
	}
```

**Comment by Narimm** [comment]
Thats probably a fix right there

**Comment by seeseemelk** [comment]
> For me this is happening when I cancel the task since the cancel operation does nothing with the `asyncTasksQueued`

This issue should be fixed by #183. Can you test this for me?

**Comment by Narimm** [comment]
Fixed
Ben Charlton

**Comment by seeseemelk** [comment]
Closing as it has been fixed.

**Comment by seeseemelk** [comment]
Reopening as I just noticed that #183 has not yet been merged.

**Comment by Narimm** [comment]
Closed by #186
